### PR TITLE
More work towards making the index pluggable

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/default_index_store.rs [hunk 1]
```diff
@@ -266,3 +274,2390 @@ fn topo_order_earlier_first(
...
+        );
+        // Merge commit and other commit
+        assert_eq!(
+            index.heads(&mut [id_5.clone(), id_3.clone()].iter()),
+            vec![id_3, id_5]
+        );
+    }
+
+    #[test]
+    fn test_hex_prefix_prefixes() {
```
Comment (yuja) [inline lib/src/default_index_store.rs:2676]: nit: these two tests don't depend on the index impl, so better to not move.

File: src/commands/mod.rs [hunk 2]
```diff
@@ -1872,11 +1871,12 @@ fn cmd_duplicate(
 
     let mut tx = workspace_command
         .start_transaction(&format!("duplicating {} commit(s)", to_duplicate.len()));
-    let index = tx.base_repo().readonly_index().clone();
-    let store = tx.base_repo().store().clone();
+    let base_repo = tx.base_repo().clone();
+    let store = base_repo.store().clone();
```
Comment (yuja) [inline src/commands/mod.rs:1875]: nit: no more need to `.clone()` the store

File: lib/src/default_index_store.rs [hunk 3]
```diff
@@ -192,9 +192,17 @@ impl IndexStore for DefaultIndexStore {
...
+        index: Box<dyn MutableIndex>,
         op_id: &OperationId,
     ) -> Result<Box<dyn ReadonlyIndex>, IndexWriteError> {
+        let index = index
+            .as_any()
+            .downcast_ref::<MutableIndexImpl>()
+            .expect("index to merge in must be a MutableIndexImpl");
+        // TODO: This clone should not be necessary. Maybe we won't need it once trait
+        // upcasting is possible. Then maybe we can upcast Box<dyn MutableIndex> to
+        // Box<dyn Any> and then downcast it to Box<MutableIndexImpl>.
```
Comment (yuja) [inline lib/src/default_index_store.rs:204]: `fn into_any(self: Box<Self>) -> Box<dyn Any>` can be added to `MutableIndex`. Then, we can downcast it to `Box<MutableIndexImpl>`.
https://doc.rust-lang.org/std/any/trait.Any.html#method.downcast

**Review by yuja — APPROVED** [review]

File: lib/src/default_index_store.rs (hunk 1 above)
Comment (martinvonz) [reply lib/src/default_index_store.rs]: Good catch. Fixed.

File: src/commands/mod.rs (hunk 2 above)
Comment (martinvonz) [reply src/commands/mod.rs]: Fixed.

File: lib/src/default_index_store.rs (hunk 3 above)
Comment (martinvonz) [reply lib/src/default_index_store.rs]: Nice! I had somehow thought that that would make the trait not object safe, but it makes sense that it still is object safe.

**Review by martinvonz — COMMENTED** [review]
Thanks for reviewing!

**Pull request merged**